x-pack/lib/monitoring/inputs/metrics.rb

```diff
@@ -211,9 +211,18 @@ def fetch_global_stats
     end
 
     def extract_cluster_uuids(stats)
-      result = stats.extract_metrics([:stats, :pipelines, :main, :config], :cluster_uuids)
-      if result && !result[:cluster_uuids].empty?
-        cluster_uuids = result[:cluster_uuids]
+      cluster_uuids = []
+      agent.running_pipelines.each do |pipeline_id, _|
+        unless pipeline_id.to_sym == :".monitoring-logstash"
+          path = [:stats, :pipelines, pipeline_id.to_sym, :config]
+          found_cluster_uuids = stats.extract_metrics(path, :cluster_uuids)
+          if found_cluster_uuids && !found_cluster_uuids[:cluster_uuids].empty?
+            cluster_uuids |= found_cluster_uuids[:cluster_uuids]
+          end
+        end
+      end
+
+      unless cluster_uuids.empty?
         @logger.info("Found cluster_uuids from elasticsearch output plugins", :cluster_uuids => cluster_uuids)
         if LogStash::SETTINGS.set?("monitoring.cluster_uuid")
           @logger.warn("Found monitoring.cluster_uuid setting configured in logstash.yml while using the ones discovered from elasticsearch output plugins, ignoring setting monitoring.cluster_uuid")
```

Comment:
this is outside the diff, but rather surprising to me:

When a user explicitly tells us that they want to monitor a cluster with `monitoring.cluster_uuid`, and at least one of their pipelines connects to an elasticsearch instance, we ignore the explicitly-given value in favor of a detected one?

Reply:
We had this discussion [here](https://github.com/elastic/logstash/pull/11541#discussion_r382258928), as it was also surprising to me, but we discussed this with @ycombinator, and we decided to go the same route as the beats team did here to keep the implementation consistent.